### PR TITLE
feat(skore/EstimatorReport)!: Move arguments from `.summarize()` to `.frame()`

### LINKED ISSUE #2533: EstimatorReport: Move arguments from `.summarize()` to `.frame()`
### What would you like to say?

Arguments of `.summarize()` that only affect the displaying of information (e.g. `flat_index` and `favorability`) should be moved to `MetricsSummaryDisplay.frame()`.

### PR DESCRIPTION
**Pull request by auguste-probabl**
#### Change description

`EstimatorReport.metrics.summarize()` no longer accepts arguments `flat_index` or `favorability`. These have been moved to `MetricsSummaryDisplay.frame()`.
This moves the responsibility of displaying things from `EstimatorReport._MetricsAccessor` to `MetricsSummaryDisplay`. 

```python
# Before
report.metrics.summarize(flat_index=True, favorability=True).frame()

# After
report.metrics.summarize().frame(flat_index=True, favorability=True)
```

This is a breaking change.

The rest of the PR consists in various refactorings, in particular the tests have been updated to reflect the change in responsibility:
- Tests of `summarize()` were extracted from `tests/unit/reports/estimator/metrics/test_numeric.py` to a new file, `tests/unit/reports/estimator/metrics/test_summarize.py`.
- `test_summarize.py` now tests that `summarize()` behaves well, and that the output of `summarize()` has a well-formed DataFrame. `.frame()` is never called.
- `displays/metrics_summary/test_estimator.py` has been rewritten to specifically test `.frame()` arguments.

A number of commits are included that could be pulled out into a separate PR if needed.

Closes #2533 
Supersedes #1839 in part

#### Contribution checklist

- [x] Unit tests were added or updated (if necessary)
- [x] Documentation was added or updated (if necessary)

#### AI usage disclosure

AI tools were involved for:
- [ ] Code generation (e.g., when writing an implementation or fixing a bug)
- [x] Test/benchmark generation
- [ ] Documentation (including examples)
- [ ] Research and understanding

In particular I used Claude to increase coverage.

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py [hunk 1]
```diff
@@ -1,25 +1,126 @@
...
+            if df["data_source"].nunique() == 1:
+                df = df.drop(columns="data_source")
+            else:
+                # Show metrics one column per data source
+                df_pivoted = df.reset_index().pivot_table(
+                    index=df.index.names,
+                    columns="data_source",
+                    values=estimator_name,
+                    sort=False,
+                )
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py:101]: This crashes when `data_source="both"` and `flat_index=True`.

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 2]
```diff
@@ -255,25 +235,26 @@ class is set to the one provided when creating the report. If `None`,
                 metrics = ["_r2", "_rmse"]
             metrics += ["_fit_time", "_predict_time"]
 
-        if metric_names is None:
-            metric_names = [None] * len(metrics)  # type: ignore
+        if metric_verbose_names is None:
+            metric_verbose_names = [None] * len(metrics)  # type: ignore
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: Could we cast `metric_verbose_names` to the right type here to avoid the `type: ignore`?

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 3]
```diff
@@ -385,102 +375,140 @@ class is set to the one provided when creating the report. If `None`,
 
             score = metric_fn(data_source=data_source, X=X, y=y, **metrics_kwargs)
 
-            index: pd.Index | pd.MultiIndex | list[str] | None
-            score_array: NDArray
             if self._parent._ml_task == "binary-classification":
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: The whole section following this is quite hard to read. Could a `match .. case` block help? 
Also it looks like the assignment of `"metric"`, `"verbose_name"`, `"estimator_name"`, `"data_source"` and `"favorability"` can be factored out of the `if/else` statements.
Or maybe it would be more readable to assign the remaining keys (`"label"`, `"average"`, `"output"` and `"score"`) one by one in respective block. I.e. do:
```
if ~statement:
    for output_idx, output_score in enumerate(score):
        scores["output"].append(output_idx)
else:    
    scores["output"].append(None)
```
four times.

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 4]
```diff
@@ -385,102 +375,140 @@ class is set to the one provided when creating the report. If `None`,
...
+                scores["label"].append(None)
+                scores["average"].append(None)
+                scores["output"].append(None)
+                scores["score"].append(score)
+                scores["favorability"].append(metric_favorability)
+
+        if any(isinstance(label, int) for label in scores["label"]):
+            scores["label"] = pd.Series(scores["label"], dtype=pd.Int64Dtype()).astype(
+                "object"
             )
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: I'm confused by this type change.

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py [hunk 5]
```diff
@@ -1,25 +1,126 @@
...
             The report metrics as a dataframe.
         """
-        return self.summarize_data
+        df = self.data.copy()
+
+        if self.report_type == "estimator":
+            df = df.dropna(axis="columns", how="all")
+
+            # TODO: Is "metric" ever needed?
+            df = df.drop(columns="metric")
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py]: Looks like only `verbose_name` is used so the `metric` column probably does not have a reason to exist. Maybe store the verbose name in the `metric` column then?

File: skore/src/skore/_utils/_index.py [hunk 6]
```diff
@@ -1,41 +1,87 @@
...
+            join_str.join(filter(bool, map(str, values)))
+            .replace(" ", join_str)
             .replace("#", "")
             .lower()
             for values in index
         ]
     )
+
+
+def rename_columns_and_index(df: pd.DataFrame, mapping: dict[str, str]) -> pd.DataFrame:
```
Comment (GaetandeCast) [inline skore/src/skore/_utils/_index.py]: Is it really necessary to create a dedicated utility function here?

**Review by GaetandeCast — COMMENTED** [review]
Here is a first pass, will look at the tests in a second time.

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py (hunk 1 above)
Comment (cakedev0) [reply skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py:101]: Found this too. Here is a non-regression test:

```Python
def test_data_source_both_flat_index(forest_binary_classification_data):
    """Check that flat_index works with `data_source="both"`."""
    estimator, X, y = forest_binary_classification_data
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

    report = EstimatorReport(
        estimator, X_train=X_train, y_train=y_train, X_test=X_test, y_test=y_test
    )

    result = report.metrics.summarize(data_source="both").frame(flat_index=True)

    assert isinstance(result.index, pd.Index)
    assert result.columns.to_list() == [
        "RandomForestClassifier (train)",
        "RandomForestClassifier (test)",
    ]

```

**Comment by cakedev0** [comment]
Note: I would have avoided to move tests while also modifying them. If I can, I always first modify then move in another PR, so it's easier to review.

The second PR can be an instant review, if you say "moving [...] from [...] to [...] and renamed from `test_summarize_[...]` to `test_[...]`" I likely merge it in less than 5 minutes ^^

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py [hunk 7]
```diff
@@ -385,102 +375,140 @@ class is set to the one provided when creating the report. If `None`,
...
+                scores["verbose_name"].append(metric_verbose_name)
+                scores["estimator_name"].append(self._parent.estimator_name_)
+                scores["data_source"].append(data_source)
+                scores["label"].append(None)
+                scores["average"].append(None)
+                scores["output"].append(None)
+                scores["score"].append(score)
+                scores["favorability"].append(metric_favorability)
+
+        if any(isinstance(label, int) for label in scores["label"]):
```
Comment (cakedev0) [inline skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: `isinstance(label, int)` is `True` for booleans, so this will transform `True/False` to `1/0` (maybe it's a desired behavior?)

File: skore/src/skore/_utils/_index.py (hunk 6 above)
Comment (auguste-probabl) [reply skore/src/skore/_utils/_index.py]: It's going to be useful for the other reports too; did you mean that you'd rather put it in `metrics_summary_display.py`?

Comment (GaetandeCast) [reply skore/src/skore/_utils/_index.py]: But it's used in `metrics_summary_display.py` which is common to all reports. So I suggest in-lining it there.

**Comment by auguste-probabl** [comment]
So working through the review comments, I am becoming a bit worried that my test refactor was not a good idea: there have been several instances now where `summarize()` produces a DataFrame that is very slightly different from the ones we use for testing `.frame()`. I'll redo the refactor so that we always use `.frame()` from data that comes from a real report.

File: skore/src/skore/_utils/_index.py (hunk 6 above)
Comment (auguste-probabl) [reply skore/src/skore/_utils/_index.py]: Discussed IRL: the function will be used when implementing `.frame()` for the other report types, so I think it should exist, but it should be moved to MetricsSummaryDisplay.

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py (hunk 1 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py:101]: Well spotted; fixed

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py (hunk 2 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: done

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py (hunk 3 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: made it more compact, stopped using defaultdict

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py (hunk 4 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: added a comment

File: skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py (hunk 5 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_plot/metrics/metrics_summary_display.py]: done

File: skore/src/skore/_utils/_index.py (hunk 6 above)
Comment (auguste-probabl) [reply skore/src/skore/_utils/_index.py]: done

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py (hunk 7 above)
Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: Should be solved (added a test)

**Comment by auguste-probabl** [comment]
> Note: I would have avoided to move tests while also modifying them. If I can, I always first modify then move in another PR, so it's easier to review.

I ended up modifying tests pretty heavily, and since there are already 2 more PRs coming about CVReport and ComparisonReport I am reluctant to split this one. Nonetheless I definitely strive for good commit hygiene in PRs, so that it's easy enough to review commit-by-commit.

File: skore/src/skore/_sklearn/_estimator/metrics_accessor.py (hunk 3 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_estimator/metrics_accessor.py]: Great, it's much clearer (to me at least) now !

File: skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py [hunk 8]
```diff
@@ -109,7 +109,7 @@ def test_binary_classification(
         )
     else:
         result = getattr(report.metrics, metric_name)(data_source=data_source)
-    pd.testing.assert_frame_equal(result, expected)
+    pd.testing.assert_frame_equal(result, expected, check_index_type=False)
```
Comment (GaetandeCast) [inline skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py]: This is not needed since you modified the type in the fixture already.

File: skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py [hunk 9]
```diff
@@ -118,7 +118,7 @@ def test_binary_classification(
...
```
Comment (GaetandeCast) [inline skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py]: same as above

File: skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py [hunk 10]
```diff
@@ -173,7 +173,7 @@ def test_regression(
         )
     else:
         result = getattr(comp.metrics, metric_name)()
-    pd.testing.assert_frame_equal(result, expected)
+    pd.testing.assert_frame_equal(result, expected, check_index_type=False)
```
Comment (GaetandeCast) [inline skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py]: same as above

File: skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py [hunk 11]
```diff
@@ -209,7 +209,7 @@ def test_custom_metric_data_source_external(
         X=X_test,
         y=y_test,
     )
-    pd.testing.assert_frame_equal(result, expected)
+    pd.testing.assert_frame_equal(result, expected, check_index_type=False)
```
Comment (GaetandeCast) [inline skore/tests/unit/reports/comparison/estimator/metrics/test_numeric.py:143]: same as above

File: skore/tests/unit/utils/test_index.py [hunk 12]
```diff
@@ -57,8 +57,15 @@ def test_flatten_multi_index(input_tuples, names, expected_values):
...
-        flatten_multi_index(simple_index)
+    result = flatten_multi_index(simple_index)
+    pd.testing.assert_index_equal(result, simple_index)
+
+
+def test_flatten_multi_index_non_multi_index_with_join_str():
+    """Test that non-MultiIndex input is returned as-is regardless of join_str."""
+    simple_index = pd.Index(["a", "b"])
+    result = flatten_multi_index(simple_index, join_str="-")
+    pd.testing.assert_index_equal(result, simple_index)
```
Comment (GaetandeCast) [inline skore/tests/unit/utils/test_index.py]: parametrize into a single test ?

**Review by GaetandeCast — COMMENTED** [review]
Some nitpicks, otherwise LGTM!

File: skore/tests/unit/utils/test_index.py (hunk 12 above)
Comment (auguste-probabl) [reply skore/tests/unit/utils/test_index.py]: removed one of them

**Review by GaetandeCast — APPROVED** [review]
LGTM, thanks a lot for the effort this took!
